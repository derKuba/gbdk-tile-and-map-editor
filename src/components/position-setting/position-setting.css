:host {
  display: block;

  width: 146px;
  height: 156px;
  margin-left: 50px;

  display: flex;
  flex-direction: column;
}

.line {
  display: flex;
}

.icon-alignment {
  display: flex;
  align-items: center;
  justify-content: center;
}

.icon-alignment img {
  border-radius: 10px;
  background: #4CAF50;
  padding: 4px;
  border: 1px solid black;
}

.icon-alignment img.arrow {
  background: white;
}

.icon-alignment img:hover {
  background-color: #63e668;
}

.icon-alignment img.arrow:hover {
  background-color: #c0c0c0;
}

.box {
  width: 55px;
  height: 60px;
}

.v-pad {
  width: 36px;
  height: 60px;
}

.center {
  width: 36px;
  height: 36px;
}

.h-pad {
  width: 55px;
  height: 36px;
}

.paint-setting-icon {
  width: 25px;
  height: 25px;
}

.arrow {
  width: 32px;
  height: 32px;
}

.up {
  margin-top: 20px;
  transform: rotate(-90deg);
}
.down {
  margin-bottom: 20px;
  transform: rotate(90deg);
}
.left {
  margin-left: 20px;
  transform: rotate(180deg);
}
.right {
  margin-right: 20px;
}