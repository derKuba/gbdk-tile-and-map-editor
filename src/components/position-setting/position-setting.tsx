/* eslint-disable max-lines */
import { Component, Host, h, Prop } from '@stencil/core';
import {
    moveUp,
    moveDown,
    moveLeft,
    moveRight,
    rotateMatrix,
    flipVertically,
    flipHorizontally,
    clearField,
} from './position-setting.utils';

import Tunnel from '../../provider/StateProvider';
import { Settings } from '../../models/settingsModel';

enum DIRECTION {
    'UP' = 0,
    'DOWN' = 1,
    'LEFT' = 2,
    'RIGHT' = 3,
}

enum FLIP_TYPE {
    'VERTICALLY' = 0,
    'HORIZONTALLY' = 1,
}

@Component({
    tag: 'mvjp-position-setting',
    styleUrl: 'position-setting.css',
    shadow: true,
})
export class PositionSetting {
    /**
     *   global matrix
     */
    @Prop() matrix: number[][][];

    /**
     *   chosen matrix tile
     */
    @Prop() indexTileMatrix: number;

    /**
     *   set matrix
     */
    @Prop() setMatrix: Function;

    /**
     *   application settings
     */
    @Prop() settings: Settings;

    private moveField = direction => {
        switch (direction) {
            case DIRECTION.UP:
                moveUp(this.matrix[this.indexTileMatrix]);
                break;
            case DIRECTION.DOWN:
                moveDown(this.matrix[this.indexTileMatrix]);
                break;
            case DIRECTION.LEFT:
                moveLeft(this.matrix[this.indexTileMatrix]);
                break;
            case DIRECTION.RIGHT:
                moveRight(this.matrix[this.indexTileMatrix]);
                break;
        }
        this.matrix = [...this.matrix];
        this.setMatrix(this.matrix);
    };

    private rotateField = () => {
        this.matrix[this.indexTileMatrix] = rotateMatrix(this.matrix[this.indexTileMatrix]);
        this.matrix = [...this.matrix];
        this.setMatrix(this.matrix);
    };

    private clearField = () => {
        this.matrix[this.indexTileMatrix] = clearField(this.settings.fieldWidth, this.settings.fieldHeight);
        this.matrix = [...this.matrix];
        this.setMatrix(this.matrix);
    };

    private flipField = type => {
        switch (type) {
            case FLIP_TYPE.VERTICALLY:
                flipVertically(this.matrix[this.indexTileMatrix]);
                break;
            case FLIP_TYPE.HORIZONTALLY:
                flipHorizontally(this.matrix[this.indexTileMatrix]);
                break;
        }
        this.matrix = [...this.matrix];
        this.setMatrix(this.matrix);
    };

    render() {
        return (
            <Host>
                <div class="line">
                    <div class="box icon-alignment">
                        <img
                            class="paint-setting-icon"
                            onClick={() => this.flipField(FLIP_TYPE.HORIZONTALLY)}
                            src={'./assets/icon/freepik_h_flip.png'}
                            alt="horizontal flip icon"
                            title="Flip Matrix horizontally"
                        />
                    </div>
                    <div class="v-pad icon-alignment">
                        <img
                            class="arrow up"
                            onClick={() => this.moveField(DIRECTION.UP)}
                            src={'./assets/icon/freepik_play.png'}
                            alt="arrow up icon"
                            title="Move whole Matrix up"
                        />
                    </div>
                    <div class="box icon-alignment">
                        <img
                            class="paint-setting-icon"
                            onClick={() => this.flipField(FLIP_TYPE.VERTICALLY)}
                            src={'./assets/icon/freepik_v_flip.png'}
                            alt="flip vertically icon"
                            title="Flip Matrix vertically"
                        />
                    </div>
                </div>

                <div class="line">
                    <div class="h-pad icon-alignment">
                        <img
                            class="arrow left"
                            onClick={() => this.moveField(DIRECTION.LEFT)}
                            src={'./assets/icon/freepik_play.png'}
                            alt="arrow left icon"
                            title="Move whole Matrix left"
                        />
                    </div>
                    <div class="center"></div>
                    <div class="h-pad icon-alignment">
                        <img
                            class="arrow right"
                            onClick={() => this.moveField(DIRECTION.RIGHT)}
                            src={'./assets/icon/freepik_play.png'}
                            alt="arrow right icon"
                            title="Move whole Matrix right"
                        />
                    </div>
                </div>

                <div class="line">
                    <div class="box icon-alignment">
                        <img
                            class="paint-setting-icon"
                            onClick={() => this.rotateField()}
                            src={'./assets/icon/freepik_reload.png'}
                            alt="arrow turn around icon"
                            title="Turn matrix counter clockwise"
                        />
                    </div>
                    <div class="v-pad icon-alignment">
                        <img
                            class="arrow down"
                            onClick={() => this.moveField(DIRECTION.DOWN)}
                            src={'./assets/icon/freepik_play.png'}
                            alt="arrow down icon"
                            title="Move whole Matrix down"
                        />
                    </div>
                    <div class="box icon-alignment">
                        <img
                            class="paint-setting-icon"
                            onClick={() => this.clearField()}
                            src={'./assets/icon/freepik_trash.png'}
                            alt="arrow trash icon"
                            title="Empty Matrix"
                        />
                    </div>
                </div>
            </Host>
        );
    }
}

Tunnel.injectProps(PositionSetting, ['matrix', 'setMatrix', 'indexTileMatrix', 'settings']); // eslint-disable-line @stencil/ban-side-effects
