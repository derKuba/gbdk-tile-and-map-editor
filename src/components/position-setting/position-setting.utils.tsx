export const moveUp = (field: number[][] | number[]): number[][] | number[] => {
    const tmp: any = field[0];
    field.splice(0, 1);
    field.push(tmp);

    return field;
};

export const moveDown = (field: number[][]): number[][] => {
    const tmp = field[field.length - 1];
    field.pop();
    field.unshift(tmp);

    return field;
};

export const moveLeft = (field: number[][]): number[][] => {
    field.forEach((row: any) => {
        row = moveUp(row);
    });

    return field;
};

export const moveRight = (field: number[][]): number[][] => {
    field.forEach((row: any) => {
        row = moveDown(row);
    });

    return field;
};

// special thx to https://gist.github.com/joshwcomeau/5b600d561f07f21b1618e1ec0fde3c2e#file-snail-flip-matrix-js
const flipMatrix = matrix => matrix[0].map((_, index) => matrix.map(row => row[index]));

export const rotateMatrix = matrix => flipMatrix(matrix).reverse();
// clockwise  flipMatrix(matrix.reverse())

// special thx to https://gist.github.com/RoryMearns/cd0e8ea604c9832b3727
export const flipHorizontally = (field: number[][]): number[][] => {
    let tmp: number[] = [];
    for (let row = 0; row < field.length / 2; row++) {
        tmp = field[field.length - row - 1];
        field[field.length - row - 1] = field[row];
        field[row] = tmp;
    }
    return field;
};
// special thx to https://gist.github.com/RoryMearns/cd0e8ea604c9832b3727
export const flipVertically = (field: number[][]): number[][] => {
    for (let row = 0; row < field.length; row++) {
        for (let col = 0; col < field[0].length / 2; col++) {
            let tmp: number = field[row][field[0].length - col - 1];
            field[row][field[0].length - col - 1] = field[row][col];
            field[row][col] = tmp;
        }
    }
    return field;
};

export const clearField = (fieldHeight, fieldWidth) => {
    return Array.from({ length: fieldHeight }, () => Array.from({ length: fieldWidth }, () => 0));
};
