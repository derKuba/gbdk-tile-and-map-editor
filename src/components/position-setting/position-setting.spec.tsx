import { newSpecPage } from '@stencil/core/dist/testing';
import { h } from '@stencil/core';

import { PositionSetting } from './position-setting';

describe('mvjp-position-setting', () => {
    let page;
    beforeEach(async () => {
        const matrix = [];
        matrix[0] = [
            [1, 2, 3],
            [4, 5, 6],
            [7, 8, 9],
        ];
        page = await newSpecPage({
            components: [PositionSetting],
            template: () => {
                return (
                    <mvjp-position-setting
                        matrix={matrix}
                        indexTileMatrix={0}
                        setMatrix={() => {}}
                        settings={{
                            fieldWidth: 8,
                            fieldHeight: 8,
                        }}
                    />
                );
            },
        });
    });
    it('builds', () => {
        expect(new PositionSetting()).toBeTruthy();
    });

    it('matches snapshot', async () => {
        expect(page.root).toMatchSnapshot();
    });

    it('should move up', async () => {
        page.root.shadowRoot.querySelectorAll('img')[1].click();
        await page.waitForChanges();
        expect(page.rootInstance.matrix[0]).toEqual([
            [4, 5, 6],
            [7, 8, 9],
            [1, 2, 3],
        ]);
    });

    it('should move down', async () => {
        page.root.shadowRoot.querySelectorAll('img')[6].click();
        await page.waitForChanges();
        expect(page.rootInstance.matrix[0]).toEqual([
            [7, 8, 9],
            [1, 2, 3],
            [4, 5, 6],
        ]);
    });

    it('should move left ', async () => {
        page.root.shadowRoot.querySelectorAll('img')[3].click();
        await page.waitForChanges();
        expect(page.rootInstance.matrix[0]).toEqual([
            [2, 3, 1],
            [5, 6, 4],
            [8, 9, 7],
        ]);
    });

    it('should move right ', async () => {
        page.root.shadowRoot.querySelectorAll('img')[4].click();
        await page.waitForChanges();
        expect(page.rootInstance.matrix[0]).toEqual([
            [3, 1, 2],
            [6, 4, 5],
            [9, 7, 8],
        ]);
    });

    it('should clear field ', async () => {
        page.root.shadowRoot.querySelectorAll('img')[7].click();
        await page.waitForChanges();
        expect(page.rootInstance.matrix[0]).toEqual([
            [0, 0, 0, 0, 0, 0, 0, 0],
            [0, 0, 0, 0, 0, 0, 0, 0],
            [0, 0, 0, 0, 0, 0, 0, 0],
            [0, 0, 0, 0, 0, 0, 0, 0],
            [0, 0, 0, 0, 0, 0, 0, 0],
            [0, 0, 0, 0, 0, 0, 0, 0],
            [0, 0, 0, 0, 0, 0, 0, 0],
            [0, 0, 0, 0, 0, 0, 0, 0],
        ]);
    });

    it('should rotate counter clockwise field ', async () => {
        page.root.shadowRoot.querySelectorAll('img')[5].click();
        await page.waitForChanges();
        expect(page.rootInstance.matrix[0]).toEqual([
            [3, 6, 9],
            [2, 5, 8],
            [1, 4, 7],
        ]);
    });

    it('should flip horizontally ', async () => {
        page.root.shadowRoot.querySelectorAll('img')[0].click();
        await page.waitForChanges();
        expect(page.rootInstance.matrix[0]).toEqual([
            [7, 8, 9],
            [4, 5, 6],
            [1, 2, 3],
        ]);
    });

    it('should flip vertically ', async () => {
        page.root.shadowRoot.querySelectorAll('img')[2].click();
        await page.waitForChanges();
        expect(page.rootInstance.matrix[0]).toEqual([
            [3, 2, 1],
            [6, 5, 4],
            [9, 8, 7],
        ]);
    });
});
