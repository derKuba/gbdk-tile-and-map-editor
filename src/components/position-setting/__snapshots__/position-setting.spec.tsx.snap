// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mvjp-position-setting matches snapshot 1`] = `
<mvjp-position-setting>
  <mock:shadow-root>
    <div class="line">
      <div class="box icon-alignment">
        <img alt="horizontal flip icon" class="paint-setting-icon" src="./assets/icon/freepik_h_flip.png" title="Flip Matrix horizontally">
      </div>
      <div class="icon-alignment v-pad">
        <img alt="arrow up icon" class="arrow up" src="./assets/icon/freepik_play.png" title="Move whole Matrix up">
      </div>
      <div class="box icon-alignment">
        <img alt="flip vertically icon" class="paint-setting-icon" src="./assets/icon/freepik_v_flip.png" title="Flip Matrix vertically">
      </div>
    </div>
    <div class="line">
      <div class="h-pad icon-alignment">
        <img alt="arrow left icon" class="arrow left" src="./assets/icon/freepik_play.png" title="Move whole Matrix left">
      </div>
      <div class="center"></div>
      <div class="h-pad icon-alignment">
        <img alt="arrow right icon" class="arrow right" src="./assets/icon/freepik_play.png" title="Move whole Matrix right">
      </div>
    </div>
    <div class="line">
      <div class="box icon-alignment">
        <img alt="arrow turn around icon" class="paint-setting-icon" src="./assets/icon/freepik_reload.png" title="Turn matrix counter clockwise">
      </div>
      <div class="icon-alignment v-pad">
        <img alt="arrow down icon" class="arrow down" src="./assets/icon/freepik_play.png" title="Move whole Matrix down">
      </div>
      <div class="box icon-alignment">
        <img alt="arrow trash icon" class="paint-setting-icon" src="./assets/icon/freepik_trash.png" title="Empty Matrix">
      </div>
    </div>
  </mock:shadow-root>
</mvjp-position-setting>
`;
